cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -400,6 +490,19 @@ impl MachInstEmit for Inst {
         state: &mut EmitState,
     ) {
         let mut allocs = AllocationConsumer::new(allocs);
+
+        // Check if we need to update the vector state before emitting this instruction
+        if let Some(expected) = self.expected_vstate() {
+            if state.vstate != EmitVState::Known(expected.clone()) {
+                // Update the vector state.
+                Inst::VecSetState {
+                    rd: writable_zero_reg(),
+                    vstate: expected.clone(),
+                }
+                .emit(&[], sink, emit_info, state);
+            }
+        }
+
```

Comment:
I'm not too familiar with how islands work, so: Does the emission of this instruction need to happen after `let mut start_off = sink.cur_offset();` below?

Reply:
I think calling `emit` before that check is correct. Since we recurse into `Inst::emit` for the `vsetivli`, we pass through that check again before emitting the instruction, and then once that is emitted we check it again for the original instruction.